File: src/applications/learning.py
"""Learn manual desktop answers, with an audit trail and no credential capture."""
import json
import re
import time
from pathlib import Path
from src.applications.answers import question
from src.applications.takeover import clean_url
from src.db.database import connect
from src.setup import store as settings

EXCLUDED=re.compile(r'password|passcode|one.?time|verification.?code|security.?code|auth.?token|\botp\b|\bpin\b|signature|sign your|social security|\bssn\b|credit card|bank account|routing number',re.I)


class Learner:
    def __init__(self,worker):self.worker=worker

    def entry(self,page):
        for app_id,entry in self.worker.sessions.items():
            if page in entry.get('pages',[entry['session'].page]):return app_id,entry
        return None,None

    def enabled(self,page):
        _,entry=self.entry(page)
        return bool(entry and not entry.get('automating',True) and settings.preferences(self.worker.path).get('native_learning',True))

    def attach(self,context):
        context.expose_binding('__applyAgentLearningReady',lambda source:self.enabled(source['page']))
        context.expose_binding('__applyAgentLearn',self.receive)
        context.add_init_script(path=str(Path(__file__).with_name('learning.js')))

    def set_recording(self,entry,enabled):
        entry['automating']=not enabled
        for page in entry.get('pages',[entry['session'].page]):
            for frame in page.frames:
                try:frame.evaluate('(enabled)=>{if(window.__applyAgentSetRecording)window.__applyAgentSetRecording(enabled);else window.__applyAgentRecording=enabled}',enabled)
                except Exception:pass

    def receive(self,source,payload):
        if not self.enabled(source['page']) or not isinstance(payload,dict):return
        app_id,entry=self.entry(source['page'])
        label=payload.get('question');kind=payload.get('type');value=payload.get('value')
        if not isinstance(label,str) or not isinstance(value,str) or not label.strip() or len(label)>600 or len(value)>20000:return
        if kind not in ('text','email','tel','url','number','search','textarea','select','combobox','radio') or EXCLUDED.search(label):return
        from src.applications import store
        record=store.get(self.worker.path,app_id)
        if not record or record['status'] not in ('ready','takeover'):return
        company=record['company'] or ''
        specific=(company and company.casefold() in label.casefold()) or re.search(r'our company|this company|this (role|position)|why.*(join|work here)',label,re.I)
        scope=company.casefold() if specific else ''
        key=question(label);raw=json.dumps(value.strip())
        entry.setdefault('manual_answers',{})[key]=value.strip()
        with connect(self.worker.path) as conn:
            old=conn.execute('SELECT value FROM saved_answers WHERE question=? AND field_type=? AND company=?',(key,kind,scope)).fetchone()
            if old and old['value']==raw:return
            if not value.strip():
                if not old:return
                conn.execute('DELETE FROM saved_answers WHERE question=? AND field_type=? AND company=?',(key,kind,scope))
            else:
                conn.execute('''INSERT INTO saved_answers(question,field_type,company,value) VALUES (?,?,?,?)
                    ON CONFLICT(question,field_type,company) DO UPDATE SET value=excluded.value''',(key,kind,scope,raw))
            conn.execute('INSERT INTO learned_answers(application_id,question,field_type,value,source_url,created_at) VALUES (?,?,?,?,?,?)',
                         (app_id,key,kind,raw,clean_url(source['frame'].url),time.time()))

File: src/applications/learning.js
(() => {
  if (window.__applyAgentLearner) return;
  window.__applyAgentLearner = true;
  window.__applyAgentRecording = false;
  window.__applyAgentLearningReady().then(enabled => { window.__applyAgentRecording = enabled; }).catch(() => {});
  const excluded = /password|passcode|one.?time|verification.?code|security.?code|auth.?token|\botp\b|\bpin\b|signature|sign your|social security|\bssn\b|credit card|bank account|routing number/i;
  const label = el => [...(el.labels || [])].map(l => {
    const copy=l.cloneNode(true);copy.querySelectorAll('input,textarea,select,button').forEach(n=>n.remove());return copy.textContent;
  }).join(' ').trim() || (el.getAttribute('aria-labelledby') || '').split(/\s+/).map(id=>document.getElementById(id)?.textContent||'').join(' ').trim() || el.getAttribute('aria-label') || '';
  const pending = new Map(), dirty = new Set();
  window.__applyAgentSetRecording=enabled=>{for(const timer of pending.values())clearTimeout(timer);pending.clear();dirty.clear();window.__applyAgentRecording=enabled;};
  let combo=null;
  function save(el){
    if(!dirty.has(el) || !window.__applyAgentRecording || !el?.isConnected || !el.getClientRects().length)return;
    dirty.delete(el);
    if([...document.querySelectorAll('input[type=password]')].some(e=>e.getClientRects().length))return;
    let question=label(el), type=el.tagName==='SELECT'?'select':el.type||'text', value=el.value;
    if(['password','hidden','file','checkbox','submit','button','reset'].includes(type))return;
    if(excluded.test(question+' '+el.name+' '+el.id+' '+el.autocomplete))return;
    if(type==='radio'){
      if(!el.checked)return;
      value=question;
      const group=el.closest('fieldset,[role=radiogroup]');
      question=group?.querySelector('legend')?.textContent || group?.getAttribute('aria-label') || '';
    } else if(el.tagName==='SELECT'){
      if(el.multiple)return;
      value=el.value?el.selectedOptions[0]?.textContent||'':'';
    } else if(el.getAttribute('role')==='combobox'){
      const wrapper=el.closest('.select');
      if(!wrapper || wrapper.querySelector('.select__multi-value'))return;
      type='combobox';value=wrapper.querySelector('.select__single-value')?.textContent||'';
    }
    if(!question.trim() || excluded.test(question))return;
    window.__applyAgentLearn({question:question.trim().slice(0,600),type,value:String(value??'').slice(0,20000)}).catch(()=>{});
  }
  function schedule(el){
    if(!window.__applyAgentRecording)return;
    dirty.add(el);
    clearTimeout(pending.get(el));
    pending.set(el,setTimeout(()=>{pending.delete(el);save(el);},600));
  }
  function flush(){for(const [el,timer] of pending){clearTimeout(timer);save(el);}pending.clear();}
  document.addEventListener('focusin',e=>{if(e.isTrusted&&e.target.getAttribute?.('role')==='combobox')combo=e.target;},true);
  document.addEventListener('input',e=>{if(e.isTrusted)schedule(e.target);},true);
  document.addEventListener('change',e=>{if(e.isTrusted&&window.__applyAgentRecording&&(dirty.has(e.target)||e.target.tagName==='SELECT'||e.target.type==='radio')){dirty.add(e.target);clearTimeout(pending.get(e.target));pending.delete(e.target);save(e.target);}},true);
  document.addEventListener('blur',e=>{if(e.isTrusted){clearTimeout(pending.get(e.target));pending.delete(e.target);save(e.target);}},true);
  document.addEventListener('click',e=>{
    if(!e.isTrusted||!window.__applyAgentRecording)return;
    flush();
    if(combo && e.target.closest?.('[role=option]')){
      const selected=combo;dirty.add(selected);setTimeout(()=>save(selected),50);
    }
  },true);
  window.addEventListener('pagehide',flush);
})();
